paymentsheet/src/main/java/com/stripe/android/paymentsheet/paymentdatacollection/polling/PollingViewModel.kt

```diff
@@ -0,0 +1,188 @@
+        if (intentStatus == StripeIntent.Status.Succeeded) {
+            _uiState.update {
+                it.copy(pollingState = PollingState.Success)
+            }
+        } else {
+            _uiState.update {
+                it.copy(pollingState = PollingState.Failed)
+            }
+        }
+    }
+
+    private fun computeTimeRemaining(): Duration {
+        val originalStartTime = savedStateHandle.get<Long>(KEY_CURRENT_POLLING_START_TIME)
+
+        if (originalStartTime == null) {
+            savedStateHandle[KEY_CURRENT_POLLING_START_TIME] = timeProvider.currentTimeInMillis()
+        }
+
+        return if (originalStartTime != null) {
+            // We were polling before the process death, so let's use the previous attempt's
+            // start time.
+            val deadline = originalStartTime + args.timeLimit.inWholeMilliseconds
+            val currentTimeInMillis = timeProvider.currentTimeInMillis()
+            val millisRemaining = deadline - currentTimeInMillis
+            maxOf(millisRemaining.milliseconds, ZERO)
+        } else {
+            args.timeLimit
+        }
+    }
+
+    fun pausePolling() {
+        poller.stopPolling()
+    }
+
+    fun resumePolling() {
+        viewModelScope.launch(dispatcher) {
+            delay(args.initialDelay)
+            poller.startPolling(viewModelScope)
+        }
+    }
+
+    fun handleCancel() {
+        _uiState.update {
+            it.copy(pollingState = PollingState.Canceled)
+        }
+        poller.stopPolling()
+    }
+
+    private suspend fun observeCountdown(timeLimit: Duration) {
+        countdownFlow(timeLimit).collect { duration ->
+            _uiState.update {
+                it.copy(durationRemaining = duration)
+            }
+        }
+    }
+
+    private suspend fun observePollingResults() {
+        poller.state
+            .map { it?.toPollingState() ?: PollingState.Active }
+            .collect(this::updatePollingState)
+    }
+
+    private fun updatePollingState(pollingState: PollingState) {
+        _uiState.update {
+            it.copy(pollingState = pollingState)
+        }
+    }
+
+    data class Args(
+        val clientSecret: String,
+        val timeLimit: Duration,
+        val initialDelay: Duration,
+        val maxAttempts: Int,
+        @InjectorKey internal val injectorKey: String = DUMMY_INJECTOR_KEY
+    )
+}
+
+private fun countdownFlow(duration: Duration) = flow {
+    var remainingDuration = duration
+    emit(remainingDuration)
+
+    while (remainingDuration > ZERO) {
+        delay(1.seconds)
+        remainingDuration -= 1.seconds
+        emit(remainingDuration)
+    }
+}
+
+private fun StripeIntent.Status.toPollingState(): PollingState {
+    return when (this) {
+        StripeIntent.Status.RequiresAction -> PollingState.Active
+        StripeIntent.Status.Succeeded -> PollingState.Success
+        StripeIntent.Status.Canceled,
+        StripeIntent.Status.Processing,
```

Comment:
Does processing == failed?

Reply:
Yes, that’s the case for UPI. If we’re reusing this `PollingViewModel` for another polling payment method, we can extend this by passing in custom mappings. For instance, UPI would treat `Processing` as a failure, while another payment method would map it to `PollingState.Active`.